core/trino-main/src/main/java/io/trino/operator/TopNRankingOperator.java

```diff
@@ -291,35 +357,59 @@ public Page getOutput()
             unfinishedWork = null;
         }
 
-        if (!finishing) {
+        if (!finishing && (groupedTopNBuilder == null || !groupedTopNBuilder.isFull()) && outputIterator == null) {
             return null;
         }
 
-        if (outputIterator == null) {
+        if (outputIterator == null && groupedTopNBuilder != null) {
             // start flushing
             outputIterator = groupedTopNBuilder.buildResult();
         }
 
         Page output = null;
-        if (outputIterator.hasNext()) {
+        if (outputIterator != null && outputIterator.hasNext()) {
             // rewrite to expected column ordering
             output = outputIterator.next().getColumns(outputChannels);
         }
+        else {
+            closeGroupedTopNBuilder();
+            return null;
+        }
         updateMemoryReservation();
         return output;
     }
 
+    @Override
+    public void close()
+    {
+        closeGroupedTopNBuilder();
+    }
+
+    private void closeGroupedTopNBuilder()
+    {
+        outputIterator = null;
+        groupedTopNBuilder = null;
+        localMemoryContext.setBytes(0);
+    }
+
     @VisibleForTesting
     public int getCapacity()
     {
-        checkState(groupByHash != null);
-        return groupByHash.getCapacity();
+        return groupedTopNBuilder == null ? 0 : groupedTopNBuilder.getCapacity();
```

Comment:
This is for simplifying groupByHash ownership. So, TopNRankingOperator doesn't need to hold groupByHash ownership. The change will be even simpler when getCapacity() needs to be deleted.

Reply:
My preference is to leave this in this one class for now, since once you put something on in interface, i usually feel like you need to be a lot more clear about the expected behavior, in which case this doesn't have one. That way the messiness is contained in just this one class and not spread around.

Reply:
Then let me remove all of them and make groupedTopNBuilder and groupByHash visible for testing.